Necessary changes  to allow T0 to use T0 storage with T2 CPUs

germanfgv opened the pull request:
Fixes #10460 

#### Status
Tested wtih T0 Replay

#### Description
Set of changes necessary to allow T0 to used dedicated storage T0_CH_CERN, while using T2 computational resources. changes are:
- Add environment  section to WMTask and addEnvironmentVariable() to WMTaskHelper
--  Allows to specify environment variables to be load in worker node.
- Add setTaskEnvironmentVariables() to WMWorkload
-- Adds a method to WMWorkloadHelper that allows to set environment variables for every task in the workload. Environment variables must be provided as a dictionary.
- Add setStepCatalogOverride to WMWorkload
-- Adds a method to WMWorkloadHelper that allows to set `overrideCatalog` option for every step in the workload
- Add support for the overrideCatalog in LogCollect Tasks
-- If LogCollect is using `overrideCatalog`, then add the `--catalog` to the `edmCopyUtil` command

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
T0 PR making use of this changes: [T04566]( https://github.com/dmwm/T0/pull/4566)
JIRA ticket: https://its.cern.ch/jira/browse/CMSTZ-898

#### External dependencies / deployment changes
No

germanfgv committed:
Add methods to modify environment variables in WMTasks and overrideCatalog option in Steps. Includes unit tests for new methods.

amaltaro reviewed (changes requested):
German, I have made some comments along the code for your consideration.

In addition to that, I would like to ask you the following:
* can you please create a "New feature" GH issue specifying what is required and what the use case is (just for bookkeeping and to stick with our contribution guidelines ;))
* I believe you have also made the necessary changes to the T0 repository. Could you please link them on this PR?
* last but not least, could you please verify those functionalities through unit tests, updating and/or creating new unit tests where it's needed (at least for WMWorkload and WMTask)? Here is an example that might be a good start for you:
https://github.com/dmwm/WMCore/blob/master/test/python/WMCore_t/WMSpec_t/WMWorkload_t.py#L2051

Please let me know if you have any questions on the unit tests setup. Probably better via slack though.

germanfgv commented:
> German, I have made some comments along the code for your consideration.
> 
> In addition to that, I would like to ask you the following:
> 
> * can you please create a "New feature" GH issue specifying what is required and what the use case is (just for bookkeeping and to stick with our contribution guidelines ;))
> * I believe you have also made the necessary changes to the T0 repository. Could you please link them on this PR?
> * last but not least, could you please verify those functionalities through unit tests, updating and/or creating new unit tests where it's needed (at least for WMWorkload and WMTask)? Here is an example that might be a good start for you:
>   https://github.com/dmwm/WMCore/blob/master/test/python/WMCore_t/WMSpec_t/WMWorkload_t.py#L2051
> 
> Please let me know if you have any questions on the unit tests setup. Probably better via slack though.

@amaltaro I think I have addressed your concerns. Please let me know if there is still work to be done.

amaltaro commented on test/python/WMCore_t/WMSpec_t/WMTask_t.py:
```diff
@@ -108,6 +108,46 @@ def testAddingSteps(self):
         self.assertEqual(task1.listNodes(), ['task1', 'task2a', 'task3', 'task2b', 'task2c'])
         return
 
+    def testAddEnvironmentVariables(self):
```
Good stuff!

amaltaro reviewed (changes requested):
German, thanks for providing these changes. They are looking better now, but there are still a few changes that will make it great, comments inline ;)
Thanks for the unit tests though, they are extremely important for these critical modules!!

germanfgv commented on src/python/WMCore/WMSpec/WMTask.py:
```diff
@@ -331,6 +332,42 @@ def build(self, workingDir):
+        Used for setting overrideCatalog option for each step in the task.
+        """
+        for step in self.steps().nodeIterator():
+            step = CMSSWStepHelper(step)
```
I'm also not sure. Right now, this is adding `overrideCatalog` to every step, regardless of step type. I'm just using `CMSSWStepHelper.setOverrideCatalog` just because it already existed. I did not pay attention to the specific step type. Maybe we should migrate `setOverrideCatalog` from `CMSSWStepHelper` to `CoreHelper`?

germanfgv replied:
Sorry, I just realize this only makes sense for CMSSW steps, as it is a CMSSW option. I'll implement this change and  test that everything is ok.

germanfgv replied:
Done. I tested it and works the same. Only that now it does not create unnecessary attributes for non-CMSSW steps.

germanfgv commented on src/python/WMCore/WMSpec/WMWorkload.py:
```diff
@@ -123,6 +123,31 @@ def setStepProperties(self, assignArgs):
+            task.addEnvironmentVariables(envDict)
+        return
+
+    def setStepOverrideCatalog(self, tfcFile):
```
If that's the case, I should change both this and WMWorkload's setStepOverrideCatalog to setOverrideCatalog, as something as task.setTaskOverrideCatalog would be kind of redundant. I put Step in the name to specify that overrideCatalog is a step attribute.

germanfgv replied:
I changed them to setOverrideCatalog

germanfgv commented on src/python/WMCore/WMSpec/Steps/Executors/LogCollect.py:
```diff
@@ -158,6 +158,11 @@ def execute(self, emulator=None):
             copyCommand = "env X509_USER_PROXY=%s edmCopyUtil" % os.environ.get('X509_USER_PROXY', None)
+
+            # specify TFC if necessary
+            if hasattr(self.step.application, 'overrideCatalog'):
```
done

germanfgv commented on src/python/WMCore/WMSpec/Steps/Templates/CMSSW.py:
```diff
@@ -344,15 +344,18 @@ def setupPileup(self, pileupConfig, dbsUrl):
+        _getOverrideCatalog_
+        return the TFC specified in overrideCatalog.
+        """
+        return getattr(self.data.application,"overrideCatalog",None)
```
done!

germanfgv commented on src/python/WMCore/WMSpec/WMTask.py:
```diff
@@ -331,6 +332,42 @@ def build(self, workingDir):
+        add a key = value style setting to the environment for this
+        task
+        """
+        for key in envDict.keys():
```
done!

germanfgv commented on src/python/WMCore/WMSpec/WMTask.py:
```diff
@@ -331,6 +332,42 @@ def build(self, workingDir):
+        _addEnvironmentVariables_
+
+        add a key = value style setting to the environment for this
+        task
```
yes, I changed the text to better reflect that fact

germanfgv commented on src/python/WMCore/WMSpec/WMTask.py:
```diff
@@ -331,6 +332,42 @@ def build(self, workingDir):
+
+        Retrieve a dictionary with all environment variables defined for this task
+        """
+        envDict = self.data.environment.dictionary_()
```
done

germanfgv commented on src/python/WMCore/WMSpec/WMTask.py:
```diff
@@ -331,6 +332,42 @@ def build(self, workingDir):
+            task.setStepOverrideCatalog(tfcFile)
+        return
+
+    def getEnvironment(self):
```
done

amaltaro commented:
@germanfgv if it's ready from your side, I can then have another look and get it merged if everything looks Ok. Just let me know please (or trigger a new request review via GH interface).

amaltaro commented:
test this please

germanfgv commented on src/python/WMCore/WMSpec/WMTask.py:
```diff
@@ -331,6 +332,41 @@ def build(self, workingDir):
+        Used for setting overrideCatalog option for each step in the task.
+        """
+        for step in self.steps().nodeIterator():
+            if step.stepType == "CMSSW":
```
Due to this check, overrideCatalog is no longer being aplied to LogCollect steps. To make it work I can just add an "or" to include "WMStep" and that's it. But, probably, the best solution would involve to add setOverrideCatalog() to either `LogCollectStepHelper` or `CoreHelper` (as right now it only exists in CMSSWStepHelper). What do you think @amaltaro?

amaltaro replied:
German, I'm afraid I do not know what would be the best option here. Would you mind providing an initial commit with the best candidate you can think of? If it looks solid, then you can run further tests and proceed with that.

germanfgv replied:
@amaltaro I moved `setOverrideCatalog()` from CMSSWHelper to CoreHelper, as now it is needed in at least 2 different step types and I wanted to avoid code dups. I tested the current versión an it works as expected. So the PR is ready for you to review.

amaltaro commented:
test this please

germanfgv commented:
> test this please

I fixed a line with a Pycodestyle warning.  Everything else remains the same.

amaltaro commented:
```
    WMCore_t.MicroService_t.MSRuleCleaner_t.MSRuleCleaner_t.MSRuleCleanerTest:testPipelineAgentCont changed from success to error
    WMCore_t.Services_t.Rucio_t.Rucio_t.RucioTest:testListDataRules changed from success to error
```

Unit test errors are likely transient, related to Rucio authentication. For the first one, it would probably be a good idea to actually mock that input data. Onto the review now...

amaltaro reviewed (approved):
This is looking good to me, German.
Could you please clarify - in the initial PR description - what is the top level key to be used for these key/value overrides? I assume it's called `OverrideCatalog`, from the StdBase module:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/StdSpecs/StdBase.py

Do you think we should have another GH issue to add a validation function for this parameter? It will likely be used only by T0, so I'm happy the way things are right now as well.

amaltaro commented:
I'm going to merge it now, but feel free to leave any comments as well.
Thanks for providing and testing this new feature, German.

PS.: Oh, and I forgot one detail. For future PRs, please do separate src/* from test/* changes in different commits.

germanfgv commented:
I'm not sure I understand what your are referring to here. This [OverrideCatalog](https://github.com/dmwm/WMCore/blob/443b69b0fb3a350763249b5d11010b80c7ee39cf/src/python/WMCore/WMSpec/StdSpecs/StdBase.py#L1034) from StdBase (which is a boolean) is not the one that we are modifying. 

What we are using is the `overrideCatalog` attribute that is used on here: 
https://github.com/dmwm/WMCore/blob/e45a4425c5437d194d0d02814b7011c752c33548/src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py#L839

> This is looking good to me, German.
> Could you please clarify - in the initial PR description - what is the top level key to be used for these key/value overrides? I assume it's called `OverrideCatalog`, from the StdBase module:
> https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/StdSpecs/StdBase.py
> 
> Do you think we should have another GH issue to add a validation function for this parameter? It will likely be used only by T0, so I'm happy the way things are right now as well.

amaltaro commented:
Oh, this is my mistake then. If we have workload getters/setters that are not really associated to the workflow top level (StdBase) `OverrideCatalog` parameter, that will be source for confusion in the future.

Did you actually look into the `OverrideCatalog` workflow parameter? Was there any reason that that wouldn't suffice for this feature?

amaltaro commented:
German, can you please also point me to the required change on the T0 repository to actually use this feature?

germanfgv commented:
> German, can you please also point me to the required change on the T0 repository to actually use this feature?

It is on the _Related PRs_ section of the description:
https://github.com/dmwm/T0/pull/4566
Should I put that info in another place?

amaltaro commented:
Oh no, that's the correct place, I actually missed it :)

@germanfgv I believe your first attempt was using the `OverrideCatalog` parameter from StdBase, no? If so, could you please share your findings and why it was not enough for this use case?

I might have a suggestion that will make things more consistent and cleaner...
